Description: allow to configure the SMTP Reject reason. This patch adds the RejectString option.
URL: https://sf.net/p/opendmarc/tickets/???/
Author: M. Favero
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Index: opendmarc-1.3.1/opendmarc/opendmarc.c
===================================================================
diff -Naru opendmarc-1.3.1.orig/opendmarc/opendmarc.c opendmarc-1.3.1/opendmarc/opendmarc.c
--- opendmarc-1.3.1.orig/opendmarc/opendmarc.c	2015-02-23 21:31:51.000000000 +0100
+++ opendmarc-1.3.1/opendmarc/opendmarc.c	2016-05-12 10:24:07.285147317 +0200
@@ -167,6 +167,7 @@
 	char *			conf_historyfile;
 	char *			conf_pslist;
 	char *			conf_ignorelist;
+	char *			conf_rejectstring;
 	char **			conf_trustedauthservids;
 	char **			conf_ignoredomains;
 };
@@ -1261,6 +1262,10 @@
 		                  &conf->conf_rejectfail,
 		                  sizeof conf->conf_rejectfail);
 
+		(void) config_get(data, "RejectString",
+                                  &conf->conf_rejectstring,
+                                  sizeof conf->conf_rejectstring);
+
 		(void) config_get(data, "RequiredHeaders",
 		                  &conf->conf_reqhdrs,
 		                  sizeof conf->conf_reqhdrs);
@@ -1359,6 +1364,33 @@
 		dmarcf_init_syslog(log_facility);
 	}
 
+        if ( conf->conf_rejectstring == NULL ) {
+                conf->conf_rejectstring = DEFREJECTSTR;
+        }
+        else {
+                /* Count occurences of "%s" in RejectString */
+                int countocc = 0;
+                const char *tmp = conf->conf_rejectstring;
+                while(tmp = strstr(tmp, "%s"))
+                {
+                        countocc++;
+                        tmp++;
+                }
+                switch ( countocc ) {
+                        case 0:
+                                snprintf(err, errlen, "%s: The RejectString doesn't contain %%s!",
+                                         basedir);
+                                return -1;
+                        case 1:
+                                break;
+                        default:
+                                snprintf(err, errlen, "%s: The RejectString contains %d occurences of %%s instead of one!",
+                                                basedir, countocc);
+                                return -1;
+                }
+	}
+
+
 	return 0;
 }
 
@@ -2942,7 +2974,7 @@
 		if (conf->conf_rejectfail && random() % 100 < pct)
 		{
 			snprintf(replybuf, sizeof replybuf,
-			         "rejected by DMARC policy for %s", pdomain);
+			         conf->conf_rejectstring, pdomain);
 
 			status = dmarcf_setreply(ctx, DMARC_REJECT_SMTP,
 			                         DMARC_REJECT_ESC, replybuf);
diff -Naru opendmarc-1.3.1.orig/opendmarc/opendmarc.conf.5.in opendmarc-1.3.1/opendmarc/opendmarc.conf.5.in
--- opendmarc-1.3.1.orig/opendmarc/opendmarc.conf.5.in	2015-02-23 21:31:51.000000000 +0100
+++ opendmarc-1.3.1/opendmarc/opendmarc.conf.5.in	2016-05-12 10:35:18.343684359 +0200
@@ -219,6 +219,13 @@
 The default is "false".
 
 .TP
+.I RejectString (string)
+This string describes the reason of reject at SMTP level.
+The message MUST contain the word "%s" once, which will be replaced by
+the RFC5322.From domain.
+The default is "rejected by DMARC policy for %s"
+
+.TP
 .I ReportCommand (string)
 Indicates the shell command to which failure reports should be passed for
 delivery when
diff -Naru opendmarc-1.3.1.orig/opendmarc/opendmarc-config.h opendmarc-1.3.1/opendmarc/opendmarc-config.h
--- opendmarc-1.3.1.orig/opendmarc/opendmarc-config.h	2015-02-23 21:31:51.000000000 +0100
+++ opendmarc-1.3.1/opendmarc/opendmarc-config.h	2016-05-11 15:25:16.699394040 +0200
@@ -41,6 +41,7 @@
 	{ "RecordAllMessages",		CONFIG_TYPE_BOOLEAN,	FALSE },
 	{ "RequiredHeaders",		CONFIG_TYPE_BOOLEAN,	FALSE },
 	{ "RejectFailures",		CONFIG_TYPE_BOOLEAN,	FALSE },
+        { "RejectString",		CONFIG_TYPE_STRING,	FALSE },
 	{ "ReportCommand",		CONFIG_TYPE_STRING,	FALSE },
 	{ "Socket",			CONFIG_TYPE_STRING,	FALSE },
 	{ "SoftwareHeader",		CONFIG_TYPE_BOOLEAN,	FALSE },
diff -Naru opendmarc-1.3.1.orig/opendmarc/opendmarc.conf.sample opendmarc-1.3.1/opendmarc/opendmarc.conf.sample
--- opendmarc-1.3.1.orig/opendmarc/opendmarc.conf.sample	2015-02-23 21:31:51.000000000 +0100
+++ opendmarc-1.3.1/opendmarc/opendmarc.conf.sample	2016-05-12 10:33:58.111681687 +0200
@@ -241,6 +241,14 @@
 #
 # RejectFailures false
 
+## RejectString string
+##	default ("rejected by DMARC policy for %s")
+##
+## This string describes the reason of reject. The message MUST contain the
+## word "%s" (only once), which will be replaced by the RFC5322.From domain.
+#
+# RejectString rejected by DMARC policy for %s
+
 ##  ReportCommand string
 ##  	default "/usr/sbin/sendmail -t"
 ##
diff -Naru opendmarc-1.3.1.orig/opendmarc/opendmarc.h opendmarc-1.3.1/opendmarc/opendmarc.h
--- opendmarc-1.3.1.orig/opendmarc/opendmarc.h	2015-02-23 21:31:51.000000000 +0100
+++ opendmarc-1.3.1/opendmarc/opendmarc.h	2016-05-11 15:11:04.897439991 +0200
@@ -34,6 +34,7 @@
 #define	BUFRSZ		2048
 #define	DEFCONFFILE	CONFIG_BASE "/opendmarc.conf"
 #define	DEFREPORTCMD	"/usr/sbin/sendmail -t -odq"
+#define DEFREJECTSTR	"rejected by DMARC policy for %s"
 #define	JOBIDUNKNOWN	"(unknown-jobid)"
 #define	MAXARGV		65536
 #define	MAXHEADER	1024
